feature: string syntax

Mwexim opened the pull request:
This pull request adds some string syntax and improves the existing ones. Note that many of these syntax classes where copied from the main Skript repository. Most of them did get some changes though.
- extended syntax for ExprStringCase
- multiple indices support in ExprStringCharAt
- ExprSubString
- ExprHashedString
- tests for all these syntax classes

The EffReplaceWith will be added in a future pull request, together with the Regex type.

Commit:
improve: extended syntax for ExprStringCase

Commit:
improve: allow multiple indices in ExprStringCharAt

Commit:
fix: ClassCastException in StringUtils

Commit:
feature: ExprSubstring

Commit:
feature: ExprHashedString

Olyno commented:
In my opinion, we should not make available the md5 and sha syntax. We should make a crypto module and include all this stuff inside, with better encryptions.

Mwexim commented:
> In my opinion, we should not make available the md5 and sha syntax. We should make a crypto module and include all this stuff inside, with better encryptions.

I agree with this. I never understood why Skript added this, but I added it anyway.
Personally, I think this is some addon stuff we should not mess around with.

Should I remove it from the PR?

Olyno commented:
> Should I remove it from the PR?

It's a yes for me, what about you @Syst3ms?

Syst3ms commented on src/main/java/io/github/syst3ms/skriptparser/expressions/ExprHashedString.java:
```diff
@@ -0,0 +1,79 @@
+ * Hashes the given text using the MD5 or SHA-256 algorithms.
+ * Each algorithm is suitable for different use cases.
+ * <ul>
+ * 	<li>MD5 is provided mostly for backwards compatibility, as it is outdated and not secure.</li>
```
I'd rather not even allow MD5 usage. I doubt the amount of Skript use cases requiring MD5 backwards-compatibility is very high, and new implementations should be encouraged to use SHA-256. If someone wants to use MD5, they'll have to make an addon, let's not implement that security risk ourselves.

WeeskyBDW commented on src/main/java/io/github/syst3ms/skriptparser/util/StringUtils.java:
```diff
@@ -324,22 +324,79 @@ else if (plural)
+     */
+    public static String toTitleCase(String str, boolean allWords) {
+        if (!allWords) {
+            return Character.toUpperCase(str.charAt(0)) + str.substring(1);
+        } else {
+            String[] words = str.split(" ");
+
+            for (int i = 0; i < words.length; i++) {
+                words[i] = Character.toUpperCase(words[i].charAt(0)) + words[i].substring(1);
+            }
+            return String.join(" ", words);
+        }
+    }
+
+    public static String toCamelCase(String str, boolean firstNoCase) {
         String[] parts = str.split("\\s+");
         StringBuilder ret = new StringBuilder();
         for (String part : parts) {
             if (firstNoCase) {
                 firstNoCase = false;
-                ret.append(part.toLowerCase());
+                ret.append(Character.toLowerCase(part.charAt(0)))
+                        .append(part.substring(1));
             } else {
                 ret.append(Character.toUpperCase(part.charAt(0)))
-                        .append(part.substring(1).toLowerCase());
+                        .append(part.substring(1));
             }
         }
         return ret.toString();
     }
```
If a skript user try to use a unicode caracter in a string who be converted as title case or as camel case, does Skript will return to it a Java Error ? due to non-handle of theses caracters

Mwexim replied:
I did not think of this, since I mainly copied the code from the Skript repository and enhanced it where necessary (some of the methods were there already).
I will test it but my personal guess is that it works fine.

WeeskyBDW reviewed (commented):
in addition of what @Mwexim and @Olyno say bellow, i just have 1 (stupid ?) question. Otherwise, LGTM

Commit:
request: add requests

Mwexim commented:
- Removed the ExprHashedString syntax class as I agree this is addon-stuff and has no particular use-case in Skript currently.
- Added a small test for camel case with Unicode characters (this applies to Pascal case as well since they use the same method)
